### PR TITLE
Added auto-save whitespace behavior if it changed manually

### LINKED ISSUE #13758: The button cannot remember the previous options
- Gitea version (or commit ref): 1.14.0+dev-265-g742e21aeb built with GNU Make 4.3, go1.15.5
- Git version: 2.28.0
- Operating system: macOS 11.0.1
  
  
  
- Database (use `[x]`):
  - [ ] PostgreSQL
  - [x] MySQL
  - [ ] MSSQL
  - [ ] SQLite
- Can you reproduce the bug at https://try.gitea.io:
  - [ ] Yes (provide example URL)
  - [x] No

## Description
In the interface of PR Review, I set **ignore whitespace when comparing lines**, but when I close the browser page and reopen the page, this becomes **Show all changes** again. I hope this button can stay in this state. Because now I need to reset this option every time I open this page
截屏2020-11-30 21 06 48
截屏2020-11-30 21 07 03

### PR DESCRIPTION
**Pull request by POPSuL**
Fix #13758, #9272

Works like "split/unified view" option, stores in database only for authenticated users when they changes that option.

### CODE REVIEW

File: models/migrations/v180.go [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+
+package migrations
+
+import (
+	"xorm.io/xorm"
+)
+
+func addWhitespaceBehaviorUserColumn(x *xorm.Engine) error {
+	type User struct {
+		WhitespaceBehavior bool `xorm:"NOT NULL DEFAULT false"`
```
Comment (lafriks) [inline models/migrations/v180.go]: Why is this different than in actual user struct?

Comment (POPSuL) [reply models/migrations/v180.go]: @lafriks its like https://github.com/go-gitea/gitea/blob/master/models/migrations/v141.go
And i think, sync full actual struct could possibly corrupt existing data (im not sure, but).

**Comment by POPSuL** [comment]
Idk how to fix it. Database scheme for integration tests is obsolete. 
At least https://github.com/go-gitea/gitea/blob/master/integrations/migration-test/gitea-v1.7.0.postgres.sql.gz is obsolete. That dump have no at least 3 columns.

1) why pg testing only on arm64? why not amd64 or else?
2) how i can solve that issue?

File: models/migrations/migrations.go [hunk 2]
```diff
@@ -889,3 +891,13 @@ func modifyColumn(x *xorm.Engine, tableName string, col *schemas.Column) error {
 	}
 	return nil
 }
+
+// addColumn adds new column to the specified table
+func addColumn(x *xorm.Engine, tableName string, col *schemas.Column) error {
```
Comment (lunny) [inline models/migrations/migrations.go]: But this will not check if the column is exist.

Comment (POPSuL) [reply models/migrations/migrations.go]: If column exists xorm throw an error. I think, is overkill for checking there.

**Review by techknowlogick — DISMISSED** [review]
Thanks for the PR. I'm adding a blocking review as I believe there was a proposal in a separate issue to add a KV table for user settings (or repo settings?) as then that would keep us from having to add a column to tables for each setting, and I think that should be the path we proceed with.

**Comment by POPSuL** [comment]
@techknowlogick Its good idea! But, maybe not KV (like `user_id|property|value`), but json like `user_id|ALL_PROPS_IN_JSON`?
Its more flexible and faster. Take one row with all user options is faster than fetch all possible rows.

**Comment by silverwind** [comment]
A simple JSON object in the db for ui-related settings would be awesome, yes.

**Comment by wxiaoguang** [comment]
The user setting table is ready. We can continue without adding columns to user table now. 😊

@POPSuL do you have time to work on it? Thank you for your contribution!

**Comment by POPSuL** [comment]
@wxiaoguang yes, i can. I will do it next week.

**Comment by POPSuL** [comment]
Rewritten to store in user_settings

**Comment by wxiaoguang** [comment]
I did some refactoring, make it more simple and clear

**Review by wxiaoguang — APPROVED** [review]

File: routers/web/repo/middlewares.go [hunk 3]
```diff
@@ -63,11 +63,24 @@ func SetDiffViewStyle(ctx *context.Context) {
...
 func SetWhitespaceBehavior(ctx *context.Context) {
+	const defaultWhitespaceBehavior = "show-all"
 	whitespaceBehavior := ctx.FormString("whitespace")
+	if ctx.IsSigned {
+		userWhitespaceBehavior, err := user_model.GetUserSetting(ctx.User.ID, user_model.SettingsKeyDiffWhitespaceBehavior, defaultWhitespaceBehavior)
+		if err == nil {
+			if whitespaceBehavior == "" {
+				whitespaceBehavior = userWhitespaceBehavior
+			} else if whitespaceBehavior != userWhitespaceBehavior {
+				_ = user_model.SetUserSetting(ctx.User.ID, user_model.SettingsKeyDiffWhitespaceBehavior, whitespaceBehavior)
```
Comment (lafriks) [inline routers/web/repo/middlewares.go:80]: It should probably be validated to be only one of valid values

Comment (POPSuL) [reply routers/web/repo/middlewares.go:80]: Done

**Review by lafriks — APPROVED** [review]

**Review by wxiaoguang — APPROVED** [review]

**Pull request merged**